In [1]:
import pandas as pd
import numpy as np
import ast

movie = pd.read_csv('tmdb_5000_movies.csv')
credit = pd.read_csv('tmdb_5000_credits.csv')
movie.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [2]:
credit.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [3]:
data = movie.merge(credit, on='title')
data.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

In [4]:
data = data[['movie_id', 'title', 'overview', 'genres', 'keywords', 'cast', 'crew']]
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [5]:
data.isna().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [6]:
data.dropna(inplace=True)

In [7]:
data.duplicated().sum()

0

In [8]:
data.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

awal : [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}] <br>
akhir : [action, adventure, fantasy, science fiction]

In [9]:
def extract(obj):
    list = []
    for i in ast.literal_eval(obj):
        list.append(i['name'])
    return list

In [10]:
data['genres'] = data['genres'].apply(extract)

In [11]:
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [12]:
data['keywords'] = data['keywords'].apply(extract)

In [14]:
def cast(obj):
    batas = 0
    list = []
    for i in ast.literal_eval(obj):
        if batas < 3:
            batas+=1
            list.append(i['name'])
        else:
            break
    return list


In [15]:
data['cast'] = data['cast'].apply(cast)

In [16]:
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [17]:
def ambil_director(obj):
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            return i['name']

In [18]:
data['Director'] = data['crew'].apply(ambil_director)
data.drop('crew', axis=1, inplace=True)
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast       Director  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  James Cameron

In [19]:
data['overview'] = data['overview'].apply(lambda i: i.split())
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast       Director  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  James Cameron

In [20]:
data['keywords'] = data['keywords'].apply(lambda x : [i.replace(' ','') for i in x])
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                               cast       Director  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  James Cameron

In [21]:
data['cast'] = data['cast'].apply(lambda x : [i.replace(' ','') for i in x])
data['genres'] = data['genres'].apply(lambda x : [i.replace(' ','') for i in x])
data['Director'] = data['Director'].apply(lambda x : str(x).replace(' ',''))
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            cast      Director  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  JamesCameron

In [22]:
data['Director'] = data['Director'].apply(lambda x:[x])
data.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            cast        Director  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]

In [23]:
data['tags'] = data['overview']+data['genres']+data['keywords']+data['cast']+data['Director']
data

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     [In, the, 22nd, century,, a, paraplegic, Marin...   
1     [Captain, Barbossa,, long, believed, to, be, d...   
2     [A, cryptic, message, from, Bond’s, past, send...   
3     [Following, the, death, of, District, Attorney...   
4     [John, Carter, is, a, war-weary,, former, mili...   
...                                                 ...   
4804  [El, Mariachi, just, wants, to, play, his, gui...   
4805  [A, newlywed, couple's, honeymoon, is, upended...   
4806  ["Signed,, Sealed,, Delivered", introduces, a,...   
4807  [When, ambitious, New, York, attorney, Sam, is...   
4808  [Ever, since, the, second, grade, when, he, fi...   

                                            genres  \
0     [Action, Adventure, Fantasy, ScienceFiction]   
1                     [Adventure, Fantasy, Action]   
2                       [Action, Adventure, Crime]   
3                 [Action, Crime, Drama, Thriller]   
4              [Action, Adventure, ScienceFiction]   
...                                            ...   
4804                     [Action, Crime, Thriller]   
4805                             [Comedy, Romance]   
4806             [Comedy, Drama, Romance, TVMovie]   
4807                                            []   
4808                                 [Documentary]   

                                               keywords  \
0     [cultureclash, future, spacewar, spacecolony, ...   
1     [ocean, drugabuse, exoticisland, eastindiatrad...   
2     [spy, basedonnovel, secretagent, sequel, mi6, ...   
3     [dccomics, crimefighter, terrorist, secretiden...   
4     [basedonnovel, mars, medallion, spacetravel, p...   
...                                                 ...   
4804  [unitedstates–mexicobarrier, legs, arms, paper...   
4805                                                 []   
4806  [date, loveatfirstsight, narration, investigat...   
4807                                                 []   
4808           [obsession, camcorder, crush, dreamgirl]   

                                                cast            Director  \
0      [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1         [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2          [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3          [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4        [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   
...                                              ...                 ...   
4804  [CarlosGallardo, JaimedeHoyos, PeterMarquardt]   [RobertRodriguez]   
4805       [EdwardBurns, KerryBishé, MarshaDietlein]       [EdwardBurns]   
4806         [EricMabius, KristinBooth, CrystalLowe]        [ScottSmith]   
4807          [DanielHenney, ElizaCoupe, BillPaxton]        [DanielHsia]   
4808  [DrewBarrymore, BrianHerzlinger, CoreyFeldman]   [BrianHerzlinger]   

                                                   tags  
0     [In, the, 22nd, century,, a, paraplegic, Marin...  
1     [Captain, Barbossa,, long, believed, to, be, d...  
2     [A, cryptic, message, from, Bond’s, past, send...  
3     [Following, the, death, of, District, Attorney...  
4     [John, Ca

In [24]:
film = data[['movie_id', 'title', 'tags']]
film

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     [In, the, 22nd, century,, a, paraplegic, Marin...  
1     [Captain, Barbossa,, long, believed, to, be, d...  
2     [A, cryptic, message, from, Bond’s, past, send...  
3     [Following, the, death, of, District, Attorney...  
4     [John, Carter, is, a, war-weary,, former, mili...  
...                                                 ...  
4804  [El, Mariachi, just, wants, to, play, his, gui...  
4805  [A, newlywed, couple's, honeymoon, is, upended...  
4806  ["Signed,, Sealed,, Delivered", introduces, a,...  
4807  [When, ambitious, New, York, attorney, Sam, is...  
4808  [Ever, since, the, second, grade, when, he, fi...  

[4806 rows x 3 columns]

In [25]:
film['tags'] = film['tags'].apply(lambda x : ' '.join(x))
film

C:\Users\User\AppData\Local\Temp\ipykernel_10680\145760852.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  film['tags'] = film['tags'].apply(lambda x : ' '.join(x))


movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     In the 22nd century, a paraplegic Marine is di...  
1     Captain Barbossa, long believed to be dead, ha...  
2     A cryptic message from Bond’s past sends him o...  
3     Following the death of District Attorney Harve...  
4     John Carter is a war-weary, former military ca...  
...                                                 ...  
4804  El Mariachi just wants to play his guitar and ...  
4805  A newlywed couple's honeymoon is upended by th...  
4806  "Signed, Sealed, Delivered" introduces a dedic...  
4807  When ambitious New York attorney Sam is sent t...  
4808  Ever since the second grade when he first saw ...  

[4806 rows x 3 columns]

In [27]:
film['tags'] = film['tags'].apply(lambda x: x.lower())
film.head(1)

C:\Users\User\AppData\Local\Temp\ipykernel_10680\1980630407.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  film['tags'] = film['tags'].apply(lambda x: x.lower())


movie_id   title                                               tags
0     19995  Avatar  in the 22nd century, a paraplegic marine is di...

In [28]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

def stem(text):
    y = []
    for i in text.split():
        y.append(ps.stem(i))
        
    return ' '.join(y)

In [29]:
film['tags'] = film['tags'].apply(stem)
film.tags[0]

C:\Users\User\AppData\Local\Temp\ipykernel_10680\4162945969.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  film['tags'] = film['tags'].apply(stem)


'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

In [33]:
from sklearn.feature_extraction.text import CountVectorizer 
cv = CountVectorizer(max_features=4500, stop_words='english')
vector = cv.fit_transform(film['tags']).toarray()
vector, vector.shape

(array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=int64),
 (4806, 4500))

In [34]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vector)
similarity, similarity.shape

(array([[1.        , 0.08827348, 0.08964215, ..., 0.04733811, 0.        ,
         0.        ],
        [0.08827348, 1.        , 0.06154575, ..., 0.02437575, 0.        ,
         0.02686077],
        [0.08964215, 0.06154575, 1.        , ..., 0.02475369, 0.        ,
         0.        ],
        ...,
        [0.04733811, 0.02437575, 0.02475369, ..., 1.        , 0.04222003,
         0.04321358],
        [0.        , 0.        , 0.        , ..., 0.04222003, 1.        ,
         0.09304842],
        [0.        , 0.02686077, 0.        , ..., 0.04321358, 0.09304842,
         1.        ]]),
 (4806, 4806))

In [51]:
sorted(list(enumerate(similarity[2])), reverse=True, key=lambda x: x[1])[1:6]

[(11, 0.3609974361905776),
 (1346, 0.3314563036811941),
 (29, 0.3307189138830738),
 (4077, 0.26311740579210874),
 (1749, 0.24618298195866545)]

In [63]:
def system(name_movie):
    index = film[film['title'] == name_movie].index[0]
    recomend = sorted(list(enumerate(similarity[index])), reverse=True, key=lambda x: x[1])[1:6]
    
    for i in recomend:
        print(film.iloc[i[0]].title)
        
system('Spectre')

Quantum of Solace
Never Say Never Again
Skyfall
From Russia with Love
Octopussy


In [49]:
film

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a parapleg marin is dispa...  
1     captain barbossa, long believ to be dead, ha c...  
2     a cryptic messag from bond’ past send him on a...  
3     follow the death of district attorney harvey d...  
4     john carter is a war-weary, former militari ca...  
...                                                 ...  
4804  el mariachi just want to play hi guitar and ca...  
4805  a newlyw couple' honeymoon is upend by the arr...  
4806  "signed, sealed, delivered" introduc a dedic q...  
4807  when ambiti new york attorney sam is sent to s...  
4808  ever sinc the second grade when he first saw h...  

[4806 rows x 3 columns]

In [65]:
import pickle
pickle.dump(film, open('film.pkl', 'wb'))

In [66]:
pickle.dump(similarity, open('similarity.pkl', 'wb'))